### ISSUE TITLE
[Android] NullPointerException

### ISSUE THREAD

**Issue by dhrrgn** [Microsoft/react-native-code-push#305]
I am occasionally seeing a NullPointerException crash when the app is checking for an update.  It doesn't always happen, which makes it tough to debug.  Here is a stacktrace:

```
Fatal Exception: java.lang.RuntimeException: Unable to resume activity {com.tackk.squad/com.tackk.squad.MainActivity}: java.lang.NullPointerException
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3111)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3140)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1350)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:146)
       at android.app.ActivityThread.main(ActivityThread.java:5692)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1291)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1107)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by java.lang.NullPointerException
       at com.microsoft.codepush.react.CodePush$CodePushNativeModule$5$1.onHostResume(CodePush.java:591)
       at com.facebook.react.bridge.ReactContext.onHostResume(ReactContext.java:147)
       at com.facebook.react.ReactInstanceManagerImpl.moveToResumedLifecycleState(ReactInstanceManagerImpl.java:559)
       at com.facebook.react.ReactInstanceManagerImpl.onHostResume(ReactInstanceManagerImpl.java:514)
       at com.facebook.react.ReactActivity.onResume(ReactActivity.java:164)
       at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1198)
       at android.app.Activity.performResume(Activity.java:5620)
       at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3101)
       at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3140)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1350)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:146)
       at android.app.ActivityThread.main(ActivityThread.java:5692)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1291)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1107)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

**Comment by lostintangent** [comment]
Hmm, interesting. @dhrrgn Could you paste an example of how you're calling `sync`? Also, is this crash happening when the app first starts and you call `sync` or after the app is resumed from the background?

Thanks for reporting this! We'll investigate it ASAP.

**Comment by dhrrgn** [comment]
It happens when the app first starts and we call `sync`.  Here is our code:

```
    this.setState({state: CHECKING}, () => {
      codePush.sync({ updateDialog: false }, (syncStatus) => {
        switch (syncStatus) {
          case codePush.SyncStatus.DOWNLOADING_PACKAGE:
            this.setState({state: DOWNLOADING});
            break;
          case codePush.SyncStatus.INSTALLING_UPDATE:
            this.setState({state: APPLYING});
            break;
        }
      })
        .then((syncStatus) => {
          switch (syncStatus) {
            case codePush.SyncStatus.UP_TO_DATE:
              this.setState({state: UP_TO_DATE}, this._done);
              break;
            case codePush.SyncStatus.UPDATE_INSTALLED:
              this.setState({state: DONE}, this._done);
              break;
            default:
              this.setState({state: DONE}, this._done);
          }
        })
        .catch(() => {
          this.setState({state: DONE}, this._done);
        });
    });
```

**Comment by dhrrgn** [comment]
Small note on where that code is:  That code is in an `InitialLaunch` component which is shown on first app launch after install to make sure they are up to date.  All other times (on resume or cold-start) we simply run: `codePush.sync({ updateDialog: false, installMode: codePush.InstallMode.ON_NEXT_RESUME });`

**Comment by lostintangent** [comment]
Ah OK, I think I might not what the issue is. Let me try to test out a fix and see if we can't get you unblocked ASAP. Apologies for this problem, and thanks so much for reporting it!

**Comment by geof90** [comment]
On the first app launch, will the cold-start call to sync which uses `codePush.InstallMode.ON_NEXT_RESUME` still happen?

**Comment by lostintangent** [comment]
Also, do you only see this issue on the first launch after install (via the `InitialLaunch` component)? Or do you also see it via the resume/cold-start call to `sync`?

**Comment by dhrrgn** [comment]
It only _seems_ to happen on the first launch.  To answer @geof90's question:  No, we have all calls to `sync` (for resume and cold-start) guarded with this:

```
     // This is only set to `true` if the InitialLaunch fully completes its CodePush sync.
    if (Settings.get('InitialLaunchCompleted', false)) {
      this._checkForUpdates(); // This function initiates the sync.
    }
```

**Comment by dhrrgn** [comment]
Just a little more info regarding the exception: We have seen crashes on Android 4.4.2 and 5.0.1, and funnily enough, it has only happened on Samsung devices (but I highly suspect that is just a coincidence).

Also, not sure this helps, but here is the stacktrace from the update check HTTP request thread:

```
OkHttp https://codepush.azurewebsites.net/updateCheck?deploymentKey=<REDACTED>&appVersion=1.5.0&packageHash=<REDACTED>&isCompanion=&label=v1&clientUniqueId=<REDACTED>
       at libcore.io.Posix.getaddrinfo(Posix.java)
       at libcore.io.ForwardingOs.getaddrinfo(ForwardingOs.java:61)
       at java.net.InetAddress.lookupHostByName(InetAddress.java:405)
       at java.net.InetAddress.getAllByNameImpl(InetAddress.java:236)
       at java.net.InetAddress.getAllByName(InetAddress.java:214)
       at com.squareup.okhttp.internal.Network$1.resolveInetAddresses(Network.java:29)
       at com.squareup.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:187)
       at com.squareup.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:156)
       at com.squareup.okhttp.internal.http.RouteSelector.next(RouteSelector.java:98)
       at com.squareup.okhttp.internal.http.HttpEngine.createNextConnection(HttpEngine.java:344)
       at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:327)
       at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:245)
       at com.squareup.okhttp.Call.getResponse(Call.java:267)
       at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:224)
       at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:195)
       at com.squareup.okhttp.Call.access$100(Call.java:34)
       at com.squareup.okhttp.Call$AsyncCall.execute(Call.java:162)
       at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:841)
```

**Comment by lostintangent** [comment]
@dhrrgn We believe we have addressed this issue, and it has now been released to NPM. Could you update your plugin and let us know if that resolves it for you? Apologies again for the inconvenience.

**Comment by dhrrgn** [comment]
@lostintangent I will update and publish a new release today.  I got a few more crash reports over the weekend for a separate issue I will be opening.

**Comment by dhrrgn** [comment]
@lostintangent The latest version has resolved this crash.  Thanks for the quick turn-around on this.

**Comment by lostintangent** [comment]
Yay! Thanks so much for confirming this. We'll try to get the other issue resolve ASAP!
